client/pubsub.go

```diff
@@ -3,61 +3,80 @@ package client
 import (
 	"context"
 	"encoding/json"
+	"log"
 
 	pb "github.com/dapr/go-sdk/dapr/proto/runtime/v1"
 	"github.com/pkg/errors"
 )
 
+// PublishEventOption is the type for the functional option.
+type PublishEventOption func(*pb.PublishEventRequest)
+
 // PublishEvent publishes data onto specific pubsub topic.
-func (c *GRPCClient) PublishEvent(ctx context.Context, pubsubName, topicName string, data []byte) error {
+func (c *GRPCClient) PublishEvent(ctx context.Context, pubsubName, topicName string, data interface{}, opts ...PublishEventOption) error {
 	if pubsubName == "" {
 		return errors.New("pubsubName name required")
 	}
 	if topicName == "" {
 		return errors.New("topic name required")
 	}
 
-	envelop := &pb.PublishEventRequest{
+	request := &pb.PublishEventRequest{
 		PubsubName: pubsubName,
 		Topic:      topicName,
-		Data:       data,
+	}
+	for _, opt := range opts {
+		opt(request)
+	}
+
+	if data != nil {
+		switch d := data.(type) {
+		case []byte:
+			request.Data = d
+		case string:
+			request.Data = []byte(d)
+		default:
+			var err error
+			request.DataContentType = "application/json"
+			request.Data, err = json.Marshal(d)
```

Comment:
Need absolute assign `err` on L39?

Reply:
Yes because it needs to be defined before L41. Otherwise, `request.Data, err := json.Marshal(d)` would throw an error because it can't initialize `request.Data`